### PR TITLE
Datacenter balance constraint

### PR DESCRIPTION
**Pull request by Crypto89**
This patch implements a dynamic constraint to force balancing of a job/task across multiple datacenters.

For some tasks, such as applications requiring quorum, it is desired (required) to spread them equally across all datacenters to prevent a loss of quorum when multiple allocations are running on the same DC.

Implements #517

### CODE REVIEW

File: scheduler/feasible_test.go [hunk 1]
```diff
@@ -612,6 +612,201 @@ func TestProposedAllocConstraint_TaskGroupDistinctHosts(t *testing.T) {
...
+	tg1 := &structs.TaskGroup{Name: "bar"}
+	tg2 := &structs.TaskGroup{Name: "baz"}
+
+	job := &structs.Job{
+		ID:          "foo",
+		Constraints: []*structs.Constraint{{Operand: structs.ConstraintBalance}},
+		Datacenters: []string{"dc1", "dc2"},
+		TaskGroups:  []*structs.TaskGroup{tg1, tg2},
+	}
+
```
Comment (dadgar) [inline scheduler/feasible_test.go]: I think what you want to test is adding an allocation to the proposed and seeing that it returns nothing

File: scheduler/feasible.go [hunk 2]
```diff
@@ -237,6 +259,83 @@ func (iter *ProposedAllocConstraintIterator) satisfiesDistinctHosts(option *stru
...
+			// skip allocation with DesiredStatus other than running
+			if alloc.DesiredStatus != structs.AllocDesiredStatusRun {
+				continue
+			}
+
+			balanceMap[node.Datacenter]++
+		}
+	}
+
+	min := math.MaxInt32
```
Comment (dadgar) [inline scheduler/feasible.go]: I think this logic should be detect the desired count of the task group and then divide it by the number of DCs.
For any DC that has less than that max you would return true and for any over you return false.

There is an edge case to consider that if the count is 10 and len(dcs) == 3, the code would have to detect that only one DC should be allowed to get to count 4.

There is a problem with the current implementation because imagine we have `count = 4` and we have 2 nodes in DC1, and one in DC2 and DC3. You would expect this to work but if the input was `[1, 1, 2, 3]`. It would return false on the second node in DC1.

File: scheduler/feasible_test.go [hunk 3]
```diff
@@ -612,6 +612,201 @@ func TestProposedAllocConstraint_TaskGroupDistinctHosts(t *testing.T) {
...
+	propsed.SetTaskGroup(tg1)
+	propsed.SetJob(job)
+
+	out := collectFeasible(propsed)
+
+	// Should only be able to schedule 1 tg on 2 nodes
+	if len(out) != 2 {
+		t.Fatalf("Bad: %#v", out)
+	}
+}
```
Comment (dadgar) [inline scheduler/feasible_test.go]: Can you add a test to stack_test.go.

That is where you can really test the behavior is working correctly.

**Comment by nanoz** [comment]
Interesting feature @Crypto89 !

**Comment by Crypto89** [comment]
So I've changed a couple of things here:

- The constraint is now called `balance_datacenters` and I've changed all variables/constants (and docs) to reflex this name so it should match
- The allocations are now deterministically determined abased on the number of allocations (tg counts) and the number of datacenters in the jobspec. this removes the need to calculate the entire state over and over again (see https://github.com/hashicorp/nomad/pull/2168#discussion_r98571633). any overflow allocations are assigned in order to the first x datacenters.
- Changed the tests to reflect this new behaviour

This last change does mean that is one of the datacenters dies with more allocations then the others, the overflow of allocations will not be rescheduled to the other datacenters. Since I personally think this is an edge case I doubt this will cause any problems in real-world usecases, however it is worth mentioning.

Last but not least, the testcases do test something, but probably not everything. since it requires a little bit more knowledge of the scheduler I would appreciate someone to give me some points/append this PR.

**Comment by preetapan** [comment]
Closing this because the work in 0.9 with the [spread stanza](https://www.nomadproject.io/docs/job-specification/spread.html) achieves this.
